### PR TITLE
Added slack webhook

### PR DESCRIPTION
**Pull request by egonm12**

### CODE REVIEW

File: app/models/transaction.rb [hunk 1]
```diff
@@ -58,4 +62,46 @@ def receiver_image
     receiver_id.nil? ? '/kabisa_lizard.png' : receiver.picture_url
   end
 
+  def send_slack_notification
+    notifier = Slack::Notifier.new ENV['WEBHOOK_URL']
+
+    notifier.ping(
+        :channel => '#kudo',
+        :attachments => [
```
Comment (thnukid) [inline app/models/transaction.rb]: Have a look at `rubocop gem` https://github.com/bbatsov/rubocop

In this hash you can use the new ruby hash syntax

```ruby
{ key: 'value, 'nested_key': { another_key: 'value' } }
```

rubocop will help you

File: Gemfile [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
-ruby '2.3.1'
+ruby '2.4.0'
 
 # Rails
 gem 'rails', '~> 5.0.0', '>= 5.0.0.1'
 
 gem 'dotenv-rails', group: [:development, :test]
 
+gem 'slack-notifier'
```
Comment (pascalw) [inline Gemfile]: It's best practise to specify a (semi) specific version of gems. This to prevent that the gem get's updated to a new major version when you run `bundle update`.

File: app/models/goal.rb [hunk 3]
```diff
@@ -17,6 +17,7 @@ def achieved?
 
   def achieve!
     touch(:achieved_on)
+
```
Comment (pascalw) [inline app/models/goal.rb]: Looks like you introduced some unnecessary whitespace here.

File: app/models/transaction.rb [hunk 4]
```diff
@@ -58,4 +62,46 @@ def receiver_image
     receiver_id.nil? ? '/kabisa_lizard.png' : receiver.picture_url
   end
 
+  def send_slack_notification
+    notifier = Slack::Notifier.new ENV['WEBHOOK_URL']
```
Comment (pascalw) [inline app/models/transaction.rb]: You generally want to use `ENV.fetch` to retrieve environment variables. If the env var is not set `ENV['WEBHOOK_URL`] will simply return nil whereas `ENV.fetch` will throw if the var is not set.

Also you'd want to make sure that during application boot the ENV var is fetched, to prevent the application can even boot with a missing environment variable.

File: app/models/transaction.rb [hunk 5]
```diff
@@ -58,4 +62,46 @@ def receiver_image
...
+
+    notifier.ping(
+        :channel => '#kudo',
+        :attachments => [
+            {
+                :fallback => 'New transaction',
+                :color => '#B58342',
+                :pretext => '<!channel> A new kudo transaction has been made! <https://kudos.kabisa.io/|Click here> for more details.',
```
Comment (pascalw) [inline app/models/transaction.rb]: Instead of hardcoding the url here you could use the `root_url` helper. This way it will work in any environment you might run this application.

File: app/models/transaction.rb (hunk 4 above)
Comment (pascalw) [inline app/models/transaction.rb]: Also I think you'd want to disable slack notifications in development.

File: app/models/transaction.rb [hunk 6]
```diff
@@ -58,4 +62,46 @@ def receiver_image
...
+                        :short => true
+                    },
+                    {
+                        :title => 'Kudos given for',
+                        :value => self.activity.name.capitalize,
+                        :short => true
+                    },
+                    {
+                        :title => 'Amount of Kudos',
+                        :value => '₭ ' + self.amount.to_s,
```
Comment (pascalw) [inline app/models/transaction.rb]: There's a `number_to_kudos` helper that will format the amount for you.

File: app/models/transaction.rb [hunk 7]
```diff
@@ -58,4 +62,46 @@ def receiver_image
...
+                        :short => true
+                    }
+                ],
+                :footer => 'Kabisa | ₭udos Platform | ' + self.created_at.to_s,
+                :footer_icon => 'https://pbs.twimg.com/profile_images/2203769368/kabisa_lizard_400x400.png',
```
Comment (pascalw) [inline app/models/transaction.rb]: To keep the application generic I think you should make this configurable. Same for `Kabisa` in the footer.

File: app/models/user.rb [hunk 8]
```diff
@@ -26,6 +26,7 @@ def self.from_omniauth(access_token)
 
   def to_s
     name
+    slack_name
```
Comment (pascalw) [inline app/models/user.rb]: What's the intention of this? You're now hiding the `name` output, since only the last expression will be returned

File: env.example [hunk 9]
```diff
@@ -1,2 +0,0 @@
-GOOGLE_CLIENT_ID=""
```
Comment (pascalw) [inline env.example]: Any reason to remove this file? IMO it's best practise to clearly document required environment variables.
On that note please add the slack webhook variable here as well.

File: app/models/transaction.rb (hunk 4 above)
Comment (pascalw) [inline app/models/transaction.rb]: Also please make the env var more descriptive, so maybe name it `SLACK_WEBHOOK_URL` or something like that.

File: app/models/user.rb (hunk 8 above)
Comment (thnukid) [inline app/models/user.rb]: Since ruby returns the last statement as a default, you will get `slack_name`. 

So `name` can be deleted if that is your intention (dunno if `name` is used somewhere else specific and you now breaking the app).

If you want both you can do:

```ruby
[name, slack_name].join(' ')
```

You might also want to add a test for this ;)

File: app/models/transaction.rb [hunk 10]
```diff
@@ -58,4 +62,46 @@ def receiver_image
     receiver_id.nil? ? '/kabisa_lizard.png' : receiver.picture_url
   end
 
+  def send_slack_notification
```
Comment (pascalw) [inline app/models/transaction.rb]: It would be preferable to extract this behaviour into a separate class that handles sending slack notifications, instead of putting this in the model. Functionality in the model should be restricted to things that actually concern the (domain) model.

While you're at it would also be nice to make the notification asynchronous for the following reasons:
* Users shouldn't have to wait until a notification is sent
* If the notification fails to send you don't want the transaction to rollback
* In a background job you'd have to possibility to retry sending the notification

Comment (pascalw) [inline app/models/transaction.rb]: Last I would to challenge you to write a (integration) test for this.

File: app/models/transaction.rb (hunk 4 above)
Comment (pascalw) [inline app/models/transaction.rb]: Also disable notifications in tests please

**Comment by pascalw** [comment]
Build is failing @egonm12, see https://travis-ci.org/kabisa/kudo-o-matic/builds/208885384

**Comment by pascalw** [comment]
@ariejan looks like travis is running twice? Why is that?

File: env.example (hunk 9 above)
Comment (ariejan) [inline env.example]: Second! Do not remove these fields in the `env.example` as it shows others what env variables need to be setup.

File: app/models/user.rb (hunk 8 above)
Comment (egonm12) [inline app/models/user.rb]: @pascalw Forgot to delete this, doesn't have any functionality.

File: app/models/transaction.rb (hunk 6 above)
Comment (egonm12) [inline app/models/transaction.rb]: @pascalw When writing `value: '₭ ' + number_to_kudos(transaction.amount),` it says `undefined method 'number_to_kudos' for SlackNotifications:Class`

File: app/models/transaction.rb (hunk 5 above)
Comment (egonm12) [inline app/models/transaction.rb]: @pascalw @pascalw When writing `pretext: "<!channel> A new kudo transaction has been made! <#{root_url}|Click here> for more details.",` it says `undefined local variable or method 'root_url' for SlackNotifications:Class` 

So instead of doing that I was thinking about `pretext: "<!channel> A new kudo transaction has been made! <#{ENV['ROOT_URL']}|Click here> for more details.",`

File: app/models/transaction.rb (hunk 6 above)
Comment (pascalw) [inline app/models/transaction.rb]: Helper methods are only available in views and controllers.
There are ways to include helpers into other classes, but my suggestion would be to make this a static utility method so it can be used easily in any context.

File: app/models/transaction.rb (hunk 5 above)
Comment (pascalw) [inline app/models/transaction.rb]: Same as my previous comment, tho in this case you can't define the helper yourself so your only option would be to include the routing helpers into your service:

```
include Rails.application.routes.url_helpers
```

File: Gemfile [hunk 11]
```diff
@@ -1,12 +1,16 @@
...
+ruby '2.4.0'
 
 # Rails
 gem 'rails', '~> 5.0.0', '>= 5.0.0.1'
 
 gem 'dotenv-rails', group: [:development, :test]
 
+gem 'slack-notifier', '~> 2.1.0'
+
+gem 'rubocop', '~> 0.47.1', require: false
```
Comment (pascalw) [inline Gemfile]: In the future try to keep pull requests focused and as small as possible (to aid the review process). Rubocop has nothing to do with slack notifications so that should go into a separate PR. Don't worry about it now just keep it in mind next time :-)

File: app/models/transaction.rb (hunk 10 above)
Comment (pascalw) [inline app/models/transaction.rb]: I see it's still synchronous. Are you planning to pick this up in this PR or are you going to tackle this separately? In any case you could probably use http://guides.rubyonrails.org/active_job_basics.html with the default async threaded backend. That's good enough for this since you can afford to lose a notification if there happens to be a job in the queue during deployment.

File: app/services/slack_notifications.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
+require 'slack-notifier'
+
+class SlackNotifications
+
+  def self.send_kudo_transaction(transaction)
+    return unless Rails.env == 'development'
```
Comment (pascalw) [inline app/services/slack_notifications.rb]: I don't think this is what you want? It will now only send notifications in development.

You could instead only send when the environment is not development or test, or only send when the `SLACK_WEBHOOK_URL` env var is set.

File: app/services/slack_notifications.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    notifier = Slack::Notifier.new ENV.fetch('SLACK_WEBHOOK_URL')
+
+    notifier.ping(
+        channel: '#testkudo',
```
Comment (pascalw) [inline app/services/slack_notifications.rb]: Would be nice to make this configurable via ENV as well

File: app/services/slack_notifications.rb [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+                        short: true
+                    },
+                    {
+                        title: 'Kudos given for',
+                        value: transaction.activity_name.capitalize,
+                        short: true
+                    },
+                    {
+                        title: 'Amount of Kudos',
+                        value: '₭ ' + transaction.amount.to_s,
```
Comment (pascalw) [inline app/services/slack_notifications.rb]: Please try using the helper to format kudos. Formatting kudos for display is a core concept that you don't want to scatter all of the code base.

**Comment by pascalw** [comment]
Hey @egonm12 can you update the travis CI build config so the test can run again? It's now failing with this:

```
Your Ruby version is 2.3.1, but your Gemfile specified 2.4.0
```
